contracts/source/UniswapOracle.sol

```diff
@@ -0,0 +1,73 @@
+pragma solidity 0.6.3;
+
+import { BlockVerifier } from "./BlockVerifier.sol";
+import { MerklePatritiaVerifier } from "./MerklePatritiaVerifier.sol";
+import { Rlp } from "./Rlp.sol";
+import { IUniswapV2Pair } from "./IUniswapV2Pair.sol";
+import { UQ112x112 } from "./UQ112x112.sol";
+
+contract UniswapOracle {
+	using UQ112x112 for uint224;
+
+	uint256 private constant MIN_BLOCK_COUNT = 1;
+	bytes32 public constant reserveTimestampSlotHash = keccak256(abi.encodePacked(uint256(8)));
+	bytes32 public constant price0SlotHash = keccak256(abi.encodePacked(uint256(9)));
+
+	IUniswapV2Pair public uniswapV2Pair;
+	bytes32 public uniswapV2PairHash;
+
+	constructor(IUniswapV2Pair _uniswapV2Pair) public {
+		uniswapV2Pair = _uniswapV2Pair;
+		uniswapV2PairHash = keccak256(abi.encodePacked(_uniswapV2Pair));
+	}
+
+	function getAccountStorageRoot(bytes memory historicBlock, bytes memory accountNodesRlp) public view returns (bytes32 storageRootHash, uint256 blockTimestamp) {
+		bytes32 stateRoot;
+		uint256 blockNumber;
+		(stateRoot, blockTimestamp, blockNumber) = BlockVerifier.extractStateRootAndTimestamp(historicBlock);
+		require (blockNumber < block.number - MIN_BLOCK_COUNT, "Proof does not cover enough blocks");
+		bytes memory accountDetailsBytes = MerklePatritiaVerifier.getValueFromProof(stateRoot, uniswapV2PairHash, accountNodesRlp);
+		Rlp.Item[] memory accountDetails = Rlp.toList(Rlp.toItem(accountDetailsBytes));
+		return (Rlp.toBytes32(accountDetails[2]), blockTimestamp);
+	}
+
+	// This function verifies the full block is old enough (MIN_BLOCK_COUNT), not too old (or blockhash will return 0x0)
+	// and return the proof values for the two storage slots we care about
```

Comment:
Style: Remove newline in comment.  Turn on word wrap if your monitor is too small.  😉